model/team_member.go

```diff
@@ -32,6 +32,11 @@ type TeamMemberForExport struct {
 	TeamName string
 }
 
+type TeamMembersWithErrors struct {
+	AddedMembers []*TeamMember     `json:"added_members"`
+	Errors       map[string]string `json:"errors"`
+}
```

Comment:
Even with this struct i feel like there is still ambiguity on how to parse it. How about something like:
```
type TeamMemberWithError struct {
    UserId		 string			   `json:"user_id"`
	TeamMember   *TeamMember       `json:"added_members"`
	Error        *AppError         `json:"error"`
}
```

and I see why you went with this as it's easier to return `[]*model.TeamMember` because it was in the struct.

If you were to go with the new proposal you'll have to convert `[]*TeamMemberWithError` to `[]*TeamMember` which is a pain, but the API becomes much cleaner. There are a few options to address the issues that arise from the new approach. 

1. Have `AddTeamMembers` return `([]*model.TeamMemberWithError, *model.AppError)` and convert `[]*model.TeamMemberWithError` to `[]*model.TeamMember` manually.
2. Have a duplicate of `AddTeamMembersBesteffort` that returns `([]*model.TeamMemberWithError, *model.AppError)`, then the original implementation stays unchanged.

My recommendation would be to go with 1, not too bad to convert `[]*model.TeamMemberWithError` to `[]*model.TeamMember`.
Thoughts?

Reply:
I'm sorry, but I don't see any benefit to the proposed change. How will this improve the API?

Reply:
First concern I have is that the current implementation returns a list of errors as string, which works in this particular case, but in the future, if we need to examine each error, string will not work. 

My second concern is that with the current API we can have a valid `Errors` value while having a valid `AddedMembers` value. Which is not go-idiomatic and will confuse people. With the proposed API, each `TeamMember` operation is isolated, if `TeamMemberWithError.Error == nil` only then you should use `TeamMemberWithError.TeamMember`, otherwise consider it an error. This simplifies the API and follows the go-idiomatic way of returning an error for an operation.

Reply:
@ali-farooq0  You convinced me :)
Fix pushed, also updated the open webapp PR and created a new redux PR https://github.com/mattermost/mattermost-redux/pull/1025 to handle API changes